ekncontent/ekncontent/eknc-domain.c

```diff
@@ -0,0 +1,866 @@
+                   "Manifest file does not hold a json object");
+      return FALSE;
+    }
+
+  JsonNode *shards_node = json_object_get_member (json_node_get_object (manifest_node), "shards");
+
+  if (shards_node == NULL || !JSON_NODE_HOLDS_ARRAY (shards_node))
+    {
+      g_set_error (error, EKNC_DOMAIN_ERROR, EKNC_DOMAIN_ERROR_BAD_MANIFEST,
+                   "Missing or malformed shards entry in manifest");
+      return FALSE;
+    }
+
+  if (!eknc_domain_clean_old_symlinks (self, subscription_dir, cancellable, error))
+    return FALSE;
+
+  JsonArray *shards_array = json_node_get_array (shards_node);
+  for (guint i = 0; i < json_array_get_length (shards_array); i++)
+    {
+      JsonNode *shard_node = json_array_get_element (shards_array, i);
+      if (!JSON_NODE_HOLDS_OBJECT (shard_node))
+        {
+          g_set_error (error, EKNC_DOMAIN_ERROR, EKNC_DOMAIN_ERROR_BAD_MANIFEST,
+                       "Malformed shard entry in manifest");
+          return FALSE;
+        }
+      JsonNode *path_node = json_object_get_member (json_node_get_object (shard_node), "path");
+      if (path_node == NULL || json_node_get_value_type (path_node) != G_TYPE_STRING)
+        {
+          g_set_error (error, EKNC_DOMAIN_ERROR, EKNC_DOMAIN_ERROR_BAD_MANIFEST,
+                       "Malformed shard entry in manifest");
+          return FALSE;
+        }
+      g_autoptr(GFile) bundle_shard_file = g_file_get_child (bundle_dir, json_node_get_string (path_node));
+      g_autoptr(GFile) subscription_shard_file = g_file_get_child (subscription_dir, json_node_get_string (path_node));
+
+      // Make a symlink if necessary
+      if (g_file_query_exists (bundle_shard_file, cancellable) &&
+          !g_file_query_exists (subscription_shard_file, cancellable) &&
+          !g_file_make_symbolic_link (subscription_shard_file,
+                                      g_file_get_path (bundle_shard_file),
+                                      cancellable, error))
+          return FALSE;
+      EosShardShardFile *shard = g_object_new (EOS_SHARD_TYPE_SHARD_FILE,
+                                               "path", g_file_get_path (subscription_shard_file),
+                                               NULL);
+      self->shards = g_slist_append (self->shards, shard);
+    }
+
+  if (!eknc_utils_parallel_init (self->shards, 0, cancellable, error))
+    return FALSE;
+
+  if (!eknc_domain_setup_link_tables (self, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+static void
+initable_iface_init (GInitableIface *initable_iface)
+{
+  initable_iface->init = eknc_domain_initable_init;
+}
+
+static EosShardRecord *
+eknc_domain_load_record_from_hash_sync (EkncDomain *self,
+                                        const gchar *hash)
+{
+  for (GSList *l = self->shards; l; l = l->next)
+    {
+      EosShardRecord *record = eos_shard_shard_file_find_record_by_hex_name (l->data, hash);
```

Comment:
I get a warning here because `const char *hash` is going into a `char *` function argument. Can be fixed by casting to `gchar *` but would be more properly fixed by marking this const in eos-shard.